src/modules/audio/spatializers/oculus_spatializer.c

```diff
@@ -0,0 +1,225 @@
+  // We maintain a list of sources and keep the index each source is associated with in its memo field.
+  // So that spatializers don't need to be notified of pauses and unpauses, we assign fields anew each onPlayback call.
+  int idx = *spatializerMemo;
+
+  // This source had a record, but we gave it away.
+  if (idx >= 0 && state.sources[idx].source != source) {
+    idx = *spatializerMemo = -1;
+  }
+
+  // This source doesn't have a record. If it's playing, try to assign it one.
+  // If there are no free source records, we will simply not play the sound,
+  // but if there's a record which is only playing a tail, in *that* case we will override the tail.
+  if (idx < 0 && lovrSourceIsPlaying(source)) {
+    if (state.occupiedCount < state.sourceMax) { // There's an empty slot
+      for (idx = 0; idx < state.sourceMax; idx++)
+        if (!state.sources[idx].occupied) // Claim the first unoccupied slot
+          break;
+    } else if (state.sourceCount < state.sourceMax) { // There's a slot doing a tail
+      for (idx = 0; idx < state.sourceMax; idx++)
+        if (!state.sources[idx].occupied // Claim the first unsourced slot
+         && !state.sources[idx].usedSourceThisPlayback) // Does OculusAudio allow reusing indexes within a playback? Let's guess no for now.
+          break;
+    }
+
+    if (idx >= 0) { // Successfully assigned
+      *spatializerMemo = idx;
+      state.sourceCount++;
+      state.occupiedCount++;
+      state.sources[idx].source = source;
+      state.sources[idx].occupied = true;
+      ovrAudio_ResetAudioSource(state.context, idx);
+    }
+  }
+
+  // This source has (or was just assigned) a record.
+  if (idx >= 0) {
+    uint32_t outStatus = 0;
+    state.sources[idx].usedSourceThisPlayback = true;
+
+    float position[4], orientation[4];
+    lovrSourceGetPose(source, position, orientation);
+
+    ovrAudio_SetAudioSourcePos(state.context, idx, position[0], position[1], position[2]);
+
+    ovrAudio_SpatializeMonoSourceInterleaved(state.context, idx, &outStatus, output, input);
+
+    if (!lovrSourceIsPlaying(source)) { // Source is finished
+      state.sources[idx].source = NULL;
+      *spatializerMemo = -1;
+      if (outStatus & ovrAudioSpatializationStatus_Finished) { // Source done playing, echo tailoff is done
+        state.sources[idx].occupied = false;
+      }
+    }
+    return framesOut;
+  }
+  return 0;
+}
+
+static uint32_t oculus_spatializer_tail(float *scratch, float* output, uint32_t frames) {
+  bool didAnything = false;
+  for (int idx = 0; idx < state.sourceMax; idx++) {
+    if (state.sources[idx].occupied && !state.sources[idx].usedSourceThisPlayback) {
+      uint32_t outStatus = 0;
+      if (!didAnything) {
+        didAnything = true;
+        memset(output, 0, frames*sizeof(float)*2);
+      }
+      ovrAudio_SpatializeMonoSourceInterleaved(state.context, idx, &outStatus, scratch, NULL);

```

Comment:
What are the semantics of passing in NULL for the input buffer?  I guess -- what is the tail actually doing here?

Reply:
In this scenario, the sound is over, so there is no input to pass in. However the spatialized version of the sound may be longer than the "normal" version of the sound. ovrAudio knows what sound we are talking about because we are passing in the index.

The lifetime of a OVR source looks like:
- Call "reset" on index to set up a new sound source
- One or more times, call SetSourcePos on the index and Spatialize on the index with input
- After done with input, call Spatialize again with NULL input on index until a Spatialize call returns the Finished status

Why are spatialized sounds longer than nonspatialized ones? How much longer? What does the "tail" contain? I... don't know. I thought it was reverb (see section "Finishing Reverb Tails" in header file docs), but actually there is a separate MixInSharedReverb function (which would go in the tail() call, but which we're not calling) and we haven't turned reverb on anyway. I assume if the engine has nothing to add, we'll get Finished essentially immediately.

Anyway I'll add a comment to make this clearer.

Reply:
I see that docs section now

> If late reverberation and simple box room modeling are enabled then the there
will be more output for the reverberation tail after the input sound has finished.
To ensure that the reverberation tail is not cut off, you can continue to feed the
spatialization functions with silence (e.g. NULL source data) after all of the
input data has been processed to get the rest of the output. When the tail has
finished the OutStatus will contain ovrAudioSpatializationStatus_Finished flag.

so because we didn't call ovrAudio_Enable with the LateReverberation and SimpleRoomModeling flags, the tails (presumably) don't do anything and they're just there for if we add reverb later?

Reply:
That sounds right to me.

Reply:
Maybe we should wait to do the tails until we're actually doing reverb?  There's still some stuff to figure out for that (deciding how to expose the room mesh/size, designing a common API/implementation for all spatializers).  It would also simplify the id assignment for now